### PR TITLE
5.5 WIP: amdgpu and drm-ttm fix

### PR DESCRIPTION
**Pull request by myfreeweb**
Works on Vega https://drmdb.emersion.fr/devices/5a737e5ae16c :)

upd: suspend-resume is kinda busted (`amdgpu_device_delayed_init_work_handler`, `ib ring test failed`, graphics in Firefox get corrupted on resume, and full crash is imminent after that) but that's an upstream bug, lots of reports that 5.6 fixes this

### CODE REVIEW

File: amd/amdgpu/Makefile [hunk 1]
```diff
@@ -546,6 +561,9 @@ CLEANFILES+= ${KMOD}.ko.full ${KMOD}.ko.debug
 
 CFLAGS+= -I${.CURDIR:H:H}/linuxkpi/gplv2/include
 CFLAGS+= -I${SYSDIR}/compat/linuxkpi/common/include
+.if exists(${SYSDIR}/compat/linuxkpi/common/linuxkpi.h)
```
Comment (myfreeweb) [inline amd/amdgpu/Makefile:564]: not sure where this came from o_0

File: drivers/gpu/drm/amd/amdgpu/amdgpu_ttm.c [hunk 2]
```diff
@@ -1802,7 +1802,7 @@ int amdgpu_ttm_init(struct amdgpu_device *adev)
 			       adev->need_dma32);
 #elif defined(__FreeBSD__)
 			       NULL, /* Dummy on BSD */
-				       NULL,
+			       adev->ddev->vma_offset_manager,
```
Comment (myfreeweb) [inline drivers/gpu/drm/amd/amdgpu/amdgpu_ttm.c:1805]: now there's a `WARN_ON` and `return -EINVAL` when this is null. hm why was this null?

File: drivers/gpu/drm/drm_gem_ttm_helper.c [hunk 3]
```diff
@@ -4,6 +4,11 @@
 
 #include <drm/drm_gem_ttm_helper.h>
 
+#ifdef __FreeBSD__
+/* Workaround for circular drm<->ttm dependency */
+extern __weak int ttm_bo_mmap_obj(struct vm_area_struct *vma, struct ttm_buffer_object *bo);
```
Comment (myfreeweb) [inline drivers/gpu/drm/drm_gem_ttm_helper.c:9]: I wonder how the circular dependency just works on Linux

File: amd/amdgpu/Makefile (hunk 1 above)
Comment (evadot) [reply amd/amdgpu/Makefile:564]: Can you remove this ?

File: drivers/gpu/drm/drm_gem_ttm_helper.c (hunk 3 above)
Comment (evadot) [reply drivers/gpu/drm/drm_gem_ttm_helper.c:9]: I have no idea, same problem exists with vram_helpers

File: amd/amdgpu/Makefile (hunk 1 above)
Comment (myfreeweb) [reply amd/amdgpu/Makefile]: done

File: drivers/gpu/drm/drm_gem_ttm_helper.c (hunk 3 above)
Comment (myfreeweb) [reply drivers/gpu/drm/drm_gem_ttm_helper.c:9]: Oh! it's probably a separate module, this file has `MODULE_DESCRIPTION("DRM gem ttm helpers");` at the bottom!

Well, a weak symbol seems like a good solution for avoiding extra modules...??

Comment (myfreeweb) [reply drivers/gpu/drm/drm_gem_ttm_helper.c:9]: Yeah, I'm pretty sure they don't work across modules, but this shouldn't cause any issues with amdgpu (unless just having a weak symbol in the module screws something up). `drm_gem_vram_helper.c` (which is the only consumer of `drm_gem_ttm_mmap` which is what has the dependency) is currently included in `vboxvideo/Makefile` only.

Huh, so these helpers have been included in our makefiles for the drivers themselves? Okay I guess we can just do the same for `drm_gem_ttm_helper.c`

**Comment by myfreeweb** [comment]
Working fine so far with the latest fixes, maybe the weird breakage at a distance (nvme interrupts even) was indeed caused by having the MSI disable function stubbed here (though that's only supposed to happen on unload?)

**Comment by evadot** [comment]
I've rebase 5.5-wip and added your commits, should be good now.
Please reopen if that's not the case.
(P.S. I always force push the -wip branch so you will need to checkout again).

**Pull request closed without merging**